packages/middleware-recursion-detection/src/index.ts

```diff
@@ -0,0 +1,64 @@
+import { HttpRequest } from "@aws-sdk/protocol-http";
+import {
+  AbsoluteLocation,
+  BuildHandler,
+  BuildHandlerArguments,
+  BuildHandlerOptions,
+  BuildHandlerOutput,
+  BuildMiddleware,
+  MetadataBearer,
+  Pluggable,
+} from "@aws-sdk/types";
+
+const TRACE_ID_HEADER_NAME = "X-Amzn-Trace-Id";
+const ENV_LAMBDA_FUNCTION_NAME = "AWS_LAMBDA_FUNCTION_NAME";
+const ENV_TRACE_ID = "_X_AMZN_TRACE_ID";
+
+interface PreviouslyResolved {
+  runtime: string;
+}
+
+/**
+ * Inject to trace ID to request header to detect recursion invocation in Lambda.
+ * @internal
+ */
+const recursionDetectionMiddleware = (options: PreviouslyResolved): BuildMiddleware<any, any>  => 
+  (next: BuildHandler<any, Output>): BuildHandler<any, Output> =>
+    async (args: BuildHandlerArguments<any>): Promise<BuildHandlerOutput<Output>> => {
+      const { request } = args;
+      if (
+        !HttpRequest.isInstance(request) ||
+        options.runtime !== "node" ||
+        request.headers.hasOwnProperty(TRACE_ID_HEADER_NAME)
+      ) {
+        return next(args);
+      }
+
+      const functionName = process.env[ENV_LAMBDA_FUNCTION_NAME];
+      const traceId = process.env[ENV_TRACE_ID];
+      const nonEmptyString = (str: unknown): str is string => typeof str === "string" && str.length > 0;
+      if (nonEmptyString(functionName) && nonEmptyString(traceId)) {
+        request.headers[TRACE_ID_HEADER_NAME] = traceId;
+      }
+      return next({
+        ...args,
+        request,
+      });
+    };
+}
```

Comment:
Remove additional semicolon and update formatting
```suggestion
const recursionDetectionMiddleware =
  (options: PreviouslyResolved): BuildMiddleware<any, any> =>
  (next: BuildHandler<any, Output>): BuildHandler<any, Output> =>
  async (args: BuildHandlerArguments<any>): Promise<BuildHandlerOutput<Output>> => {
    const { request } = args;
    if (
      !HttpRequest.isInstance(request) ||
      options.runtime !== "node" ||
      request.headers.hasOwnProperty(TRACE_ID_HEADER_NAME)
    ) {
      return next(args);
    }

    const functionName = process.env[ENV_LAMBDA_FUNCTION_NAME];
    const traceId = process.env[ENV_TRACE_ID];
    const nonEmptyString = (str: unknown): str is string => typeof str === "string" && str.length > 0;
    if (nonEmptyString(functionName) && nonEmptyString(traceId)) {
      request.headers[TRACE_ID_HEADER_NAME] = traceId;
    }
    return next({
      ...args,
      request,
    });
  };
```